File: src/ui/edit.ts
/**
 * 三级 · 编辑表单（对齐现版 Edit-Form）：一屏显示全部字段，选序号改单项。
 *
 * 新需求（plan §7）：密钥行默认掩码 `********`，提供「👁 显示/隐藏密钥明文」开关——
 * 仅影响本表单的**显示**，不改数据、不持久化；默认隐藏防肩窥。输入态（readValue secret）另算。
 */
import {
  buildProviderEnv,
  getProviderEnvMap,
  reconcileBuiltin,
  resolveUniqueName,
  type Provider,
  type Store,
} from '../config/store.js';
import type { Preset } from '../config/types.js';
import { T } from '../i18n/index.js';
import { pickAuth, pickBaseUrl, pickEffort, pickProvider, pickProviderUrl } from './pickers.js';
import { selectMenu } from './select.js';
import { readText, readValue } from './text.js';

interface WorkCopy {
  name: string;
  note: string;
  base: string;
  auth: 'AUTH_TOKEN' | 'API_KEY';
  token: string;
  opus: string;
  sonnet: string;
  haiku: string;
  effort: string;
}

function fromProvider(p: Provider): WorkCopy {
  const m = getProviderEnvMap(p);
  const usesApiKey = Boolean(m.ANTHROPIC_API_KEY && m.ANTHROPIC_API_KEY.trim());
  return {
    name: p.name,
    note: p.note ?? '',
    base: m.ANTHROPIC_BASE_URL ?? '',
    auth: usesApiKey ? 'API_KEY' : 'AUTH_TOKEN',
    token: (usesApiKey ? m.ANTHROPIC_API_KEY : m.ANTHROPIC_AUTH_TOKEN) ?? '',
    opus: m.ANTHROPIC_DEFAULT_OPUS_MODEL ?? '',
    sonnet: m.ANTHROPIC_DEFAULT_SONNET_MODEL ?? '',
    haiku: m.ANTHROPIC_DEFAULT_HAIKU_MODEL ?? '',
    effort: m.CLAUDE_CODE_EFFORT_LEVEL ?? '',
  };
}

/** 编辑 `prov`（就地修改）；保存返回 true，放弃返回 false。 */
export async function editForm(prov: Provider, store: Store, catalog: Preset[]): Promise<boolean> {
  const W = fromProvider(prov);
  let showSecret = false;
  let start = 0;

  for (;;) {
    const v = (x: string): string => (x === '' ? T('empty.paren') : x);
    const keyDisp = W.token === '' ? T('empty.paren') : showSecret ? W.token : '********';
    const rows: Array<{ action: string; label: string }> = [
      { action: 'provider', label: `${T('edit.field.provider')}: ${v(W.name)}` },
      { action: 'note', label: `${T('edit.field.note')}: ${v(W.note)}` },
      { action: 'base', label: `${T('edit.field.base')}: ${v(W.base)}` },
      { action: 'auth', label: `${T('edit.field.auth')}: ${W.auth}` },
      { action: 'key', label: `${T('edit.field.key')}: ${keyDisp}` },
      { action: 'opus', label: `${T('edit.field.opus')}: ${v(W.opus)}` },
      { action: 'sonnet', label: `${T('edit.field.sonnet')}: ${v(W.sonnet)}` },
      { action: 'haiku', label: `${T('edit.field.haiku')}: ${v(W.haiku)}` },
      { action: 'effort', label: `${T('edit.field.effort')}: ${v(W.effort)}` },
      { action: 'toggle', label: showSecret ? T('edit.toggleSecretHide') : T('edit.toggleSecretShow') },
      { action: 'sep', label: '' },
      { action: 'save', label: T('edit.save') },
      { action: 'discard', label: T('edit.discard') },
    ];

    const sel = await selectMenu({ title: T('edit.title'), items: rows.map((r) => r.label), start, hint: T('edit.hint') });
    if (sel < 0) return false; // Esc / q = 放弃
    start = sel;

    switch (rows[sel]?.action) {
      case 'provider': {
        const pp = await pickProvider(catalog, W.name);
        if (pp === 'custom') {
          const name = await readText(`  ${T('edit.customName')}`);
          if (name && name.trim()) W.name = name.trim();
        } else if (pp) {
          W.name = pp.name;
          W.auth = pp.auth;
          W.base = await pickProviderUrl(pp, W.base);
          if (pp.models.opus) W.opus = pp.models.opus;
          if (pp.models.sonnet) W.sonnet = pp.models.sonnet;
          if (pp.models.haiku) W.haiku = pp.models.haiku;
          if (pp.effort) W.effort = pp.effort;
        }
        break;
      }
      case 'note': {
        const note = await readText(`  ${T('edit.noteInput')}`);
        if (note === '-') W.note = '';
        else if (note && note.trim()) W.note = note.trim();
        break;
      }
      case 'base':
        W.base = await pickBaseUrl(W.base, store, catalog);
        break;
      case 'auth':
        W.auth = await pickAuth(W.auth);
        break;
      case 'key': {
        const r = await readValue(T('edit.field.key').trim(), W.token, true);
        if (r.changed) W.token = r.value;
        break;
      }
      case 'opus': {
        const r = await readValue(T('edit.field.opus').trim(), W.opus);
        if (r.changed) W.opus = r.value;
        break;
      }
      case 'sonnet': {
        const r = await readValue(T('edit.field.sonnet').trim(), W.sonnet);
        if (r.changed) W.sonnet = r.value;
        break;
      }
      case 'haiku': {
        const r = await readValue(T('edit.field.haiku').trim(), W.haiku);
        if (r.changed) W.haiku = r.value;
        break;
      }
      case 'effort':
        W.effort = await pickEffort(W.effort);
        break;
      case 'toggle':
        showSecret = !showSecret; // 仅切换显示，不改数据、不持久化
        break;
      case 'save': {
        if (W.name.trim() === '') {
          console.log(`  ${T('edit.nameEmpty')}`);
          break;
        }
        const fields: Record<string, string> = {
          ANTHROPIC_BASE_URL: W.base,
          ANTHROPIC_DEFAULT_OPUS_MODEL: W.opus,
          ANTHROPIC_DEFAULT_SONNET_MODEL: W.sonnet,
          ANTHROPIC_DEFAULT_HAIKU_MODEL: W.haiku,
          CLAUDE_CODE_EFFORT_LEVEL: W.effort,
        };
        if (W.auth === 'API_KEY') fields.ANTHROPIC_API_KEY = W.token;
        else fields.ANTHROPIC_AUTH_TOKEN = W.token;
        prov.name = resolveUniqueName(store, W.name.trim(), prov);
        prov.env = buildProviderEnv(fields);
        prov.note = W.note;
        reconcileBuiltin(prov); // [P1] 官方档被配成第三方后清掉 builtin 身份
        return true;
      }
      case 'discard':
        return false;
      default:
        break; // sep
    }
  }
}

File: internal/tui/edit.go
package tui

import (
	"fmt"
	"strings"

	"github.com/becomeless/cc-x/internal/config"
	"github.com/becomeless/cc-x/internal/i18n"
	"github.com/becomeless/cc-x/internal/presets"
)

type workCopy struct {
	name, note, base, auth, token, opus, sonnet, haiku, effort string
}

func fromProvider(p config.Provider) workCopy {
	m := config.GetProviderEnvMap(p)
	usesAPIKey := strings.TrimSpace(m["ANTHROPIC_API_KEY"]) != ""
	auth := presets.AuthToken
	token := m["ANTHROPIC_AUTH_TOKEN"]
	if usesAPIKey {
		auth = presets.AuthAPIKey
		token = m["ANTHROPIC_API_KEY"]
	}
	return workCopy{
		name: p.Name, note: p.Note,
		base:   m["ANTHROPIC_BASE_URL"],
		auth:   auth,
		token:  token,
		opus:   m["ANTHROPIC_DEFAULT_OPUS_MODEL"],
		sonnet: m["ANTHROPIC_DEFAULT_SONNET_MODEL"],
		haiku:  m["ANTHROPIC_DEFAULT_HAIKU_MODEL"],
		effort: m["CLAUDE_CODE_EFFORT_LEVEL"],
	}
}

func toggleLabel(show bool) string {
	if show {
		return i18n.T("edit.toggleSecretHide")
	}
	return i18n.T("edit.toggleSecretShow")
}

// EditForm 编辑 prov（就地修改）；保存返回 true，放弃返回 false。对应 npm 版 editForm。
// 密钥行默认掩码，「👁 显示/隐藏」仅切换本表单显示、不改数据、不持久化。
func EditForm(t *Terminal, prov *config.Provider, store *config.Store, catalog []presets.Preset) bool {
	w := fromProvider(*prov)
	showSecret := false
	start := 0

	v := func(x string) string {
		if x == "" {
			return i18n.T("empty.paren")
		}
		return x
	}

	for {
		keyDisp := i18n.T("empty.paren")
		if w.token != "" {
			if showSecret {
				keyDisp = w.token
			} else {
				keyDisp = "********"
			}
		}
		type row struct{ action, label string }
		rows := []row{
			{"provider", i18n.T("edit.field.provider") + ": " + v(w.name)},
			{"note", i18n.T("edit.field.note") + ": " + v(w.note)},
			{"base", i18n.T("edit.field.base") + ": " + v(w.base)},
			{"auth", i18n.T("edit.field.auth") + ": " + w.auth},
			{"key", i18n.T("edit.field.key") + ": " + keyDisp},
			{"opus", i18n.T("edit.field.opus") + ": " + v(w.opus)},
			{"sonnet", i18n.T("edit.field.sonnet") + ": " + v(w.sonnet)},
			{"haiku", i18n.T("edit.field.haiku") + ": " + v(w.haiku)},
			{"effort", i18n.T("edit.field.effort") + ": " + v(w.effort)},
			{"sep", ""},
			{"toggle", toggleLabel(showSecret)},
			{"sep", ""},
			{"save", i18n.T("edit.save")},
			{"discard", i18n.T("edit.discard")},
		}
		items := make([]string, len(rows))
		for i, r := range rows {
			items[i] = r.label
		}

		sel := SelectMenu(t, SelectOptions{Title: i18n.T("edit.title"), Items: items, Start: start, Hint: i18n.T("edit.hint"), NoNumber: true})
		if sel < 0 {
			return false // Esc / q = 放弃
		}
		start = sel

		switch rows[sel].action {
		case "provider":
			pp, custom := PickProvider(t, catalog, w.name)
			if custom {
				if name, ok := ReadText(t, "  "+i18n.T("edit.customName")); ok && strings.TrimSpace(name) != "" {
					w.name = strings.TrimSpace(name)
				}
			} else if pp != nil {
				w.name = pp.Name
				w.auth = pp.Auth
				w.base = PickProviderURL(t, pp, w.base)
				if pp.Models.Opus != "" {
					w.opus = pp.Models.Opus
				}
				if pp.Models.Sonnet != "" {
					w.sonnet = pp.Models.Sonnet
				}
				if pp.Models.Haiku != "" {
					w.haiku = pp.Models.Haiku
				}
				if pp.Effort != "" {
					w.effort = pp.Effort
				}
			}
		case "note":
			if note, ok := ReadText(t, "  "+i18n.T("edit.noteInput")); ok {
				if note == "-" {
					w.note = ""
				} else if strings.TrimSpace(note) != "" {
					w.note = strings.TrimSpace(note)
				}
			}
		case "base":
			w.base = PickBaseURL(t, w.base, store, catalog)
		case "auth":
			w.auth = PickAuth(t, w.auth)
		case "key":
			if ch, val := ReadValue(t, strings.TrimSpace(i18n.T("edit.field.key")), w.token, true); ch {
				w.token = val
			}
		case "opus":
			if ch, val := ReadValue(t, strings.TrimSpace(i18n.T("edit.field.opus")), w.opus, false); ch {
				w.opus = val
			}
		case "sonnet":
			if ch, val := ReadValue(t, strings.TrimSpace(i18n.T("edit.field.sonnet")), w.sonnet, false); ch {
				w.sonnet = val
			}
		case "haiku":
			if ch, val := ReadValue(t, strings.TrimSpace(i18n.T("edit.field.haiku")), w.haiku, false); ch {
				w.haiku = val
			}
		case "effort":
			w.effort = PickEffort(t, w.effort)
		case "toggle":
			showSecret = !showSecret
		case "save":
			if strings.TrimSpace(w.name) == "" {
				fmt.Printf("  %s\n", i18n.T("edit.nameEmpty"))
				continue
			}
			fields := map[string]string{
				"ANTHROPIC_BASE_URL":             w.base,
				"ANTHROPIC_DEFAULT_OPUS_MODEL":   w.opus,
				"ANTHROPIC_DEFAULT_SONNET_MODEL": w.sonnet,
				"ANTHROPIC_DEFAULT_HAIKU_MODEL":  w.haiku,
				"CLAUDE_CODE_EFFORT_LEVEL":       w.effort,
			}
			if w.auth == presets.AuthAPIKey {
				fields["ANTHROPIC_API_KEY"] = w.token
			} else {
				fields["ANTHROPIC_AUTH_TOKEN"] = w.token
			}
			prov.Name = config.ResolveUniqueName(store, strings.TrimSpace(w.name), prov)
			prov.Env = config.BuildProviderEnv(fields)
			prov.Note = w.note
			config.ReconcileBuiltin(prov) // 官方档被配成第三方后清掉 builtin 身份
			return true
		case "discard":
			return false
		}
	}
}
